pkg/otlp/model/translator/statspayload.go

```diff
@@ -0,0 +1,471 @@
+	Name           string
+	Resource       string
+	HTTPStatusCode uint32
+	Type           string
+	DBType         string
+	Synthetics     bool
+}
+
+// aggregationValue specifies the set of metrics corresponding to a certain aggregationKey.
+type aggregationValue struct {
+	Hits         uint64
+	Errors       uint64
+	Duration     uint64
+	OkSummary    []byte
+	ErrorSummary []byte
+	TopLevelHits uint64
+}
+
+// aggregations stores aggregation values (stats) grouped by their corresponding keys.
+type aggregations struct {
+	agg map[aggregationKey]*aggregationValue
+}
+
+// Value returns the aggregation value corresponding to the key found in map m.
+func (a *aggregations) Value(m pcommon.Map) *aggregationValue {
+	var sntx bool
+	if v, ok := m.Get(statsKeySynthetics); ok {
+		sntx = v.Bool()
+	}
+	key := aggregationKey{
+		Service:        getStr(m, statsKeyService),
+		Name:           getStr(m, statsKeySpanName),
+		Resource:       getStr(m, statsKeySpanResource),
+		HTTPStatusCode: uint32(getInt(m, statsKeyHTTPStatusCode)),
+		Type:           getStr(m, statsKeySpanType),
+		DBType:         getStr(m, statsKeySpanDBType),
+		Synthetics:     sntx,
+	}
+	if a.agg == nil {
+		a.agg = make(map[aggregationKey]*aggregationValue)
+	}
+	if _, ok := a.agg[key]; !ok {
+		a.agg[key] = new(aggregationValue)
+	}
+	return a.agg[key]
+}
+
+// Stats returns the set of pb.ClientGroupedStats based on all the aggregated key/value
+// pairs.
+func (a *aggregations) Stats() []pb.ClientGroupedStats {
+	cgs := make([]pb.ClientGroupedStats, 0, len(a.agg))
+	for k, v := range a.agg {
+		cgs = append(cgs, pb.ClientGroupedStats{
+			Service:        k.Service,
+			Name:           k.Name,
+			Resource:       k.Resource,
+			HTTPStatusCode: k.HTTPStatusCode,
+			Type:           k.Type,
+			DBType:         k.DBType,
+			Synthetics:     k.Synthetics,
+			Hits:           v.Hits,
+			Errors:         v.Errors,
+			Duration:       v.Duration,
+			OkSummary:      v.OkSummary,
+			ErrorSummary:   v.ErrorSummary,
+			TopLevelHits:   v.TopLevelHits,
+		})
+	}
+	return cgs
+}
+
+// UnsetHostnamePlaceholder is the string used as a hostname when the hostname can not be extracted from span attributes
+// by the processor. Upon decoding the metrics, the Translator will use its configured fallback SourceProvider to replace
+// it with the correct hostname.
+//
+// This isn't the most ideal approach to the problem, but provides the better user experience by avoiding the need to
+// duplicate the "exporter::datadog::hostname" configuration field as "processor::datadog::hostname". The hostname can
+// also not be left empty in case of failure to obtain it, because empty has special meaning. An empty hostname means
+// that we are in a Lambda environment. Thus, we must use a placeholder.
+const UnsetHostnamePlaceholder = "__unset__"
+
+// StatsPayloadToMetrics converts OTLP Metrics to an APM Stats Payload.
+func (t *Translator) StatsPayloadFromMetrics(mx pmetric.Metrics) (pb.StatsPayload, error) {
```

Comment:
Does this need to be public?

Reply:
This one not so much, but I'd keep it for symmetry. Is that a bad reason?

Reply:
If someone wants to make this public in the future, that's easy, if we want to make it private it's (potentially) harder because of any dependencies we may have introduced, so I think in principle it's good to limit our public API as much as possible. I don't want to block the PR on this, so I will leave it up to you to decide since I don't think it's very bad if it remains public.

Reply:
I've already made this private in my latest commit with the tests. Forgot to follow up here.